libraries/AP_Terrain/TerrainIO.cpp

```diff
@@ -178,8 +178,7 @@ void AP_Terrain::open_file(void)
     // create directory if need be
     if (!directory_created) {
         *p = 0;
-        mkdir(file_path, 0755);
```

Comment:
Let's not forget this is used in Linux too and mkdir should be working there. Also, as @WickedShell said, NuttX is supposed to be POSIX compliant, so it should be doing the right thing here.

Having that said, I don't think this is enough. If we don't get the directory created I think we need to have *io_failure* set to _true_ and just return.